stub/src/main/java/io/grpc/stub/ClientCalls.java

```diff
@@ -763,6 +858,153 @@ public void execute(Runnable runnable) {
     }
   }
 
+  @SuppressWarnings("serial")
+  static final class ThreadSafeThreadlessExecutor extends ConcurrentLinkedQueue<Runnable>
+      implements Executor {
+    private static final Logger log =
+        Logger.getLogger(ThreadSafeThreadlessExecutor.class.getName());
+
+    private static final Object SHUTDOWN = new Object(); // sentinel
+
+    // Set to the calling thread while it's parked, SHUTDOWN on RPC completion
+    private volatile Object waiter;
+    private final Collection<Thread> waitingThreads = new ArrayList<>();
+    private final List<ValidatingThread<?>> validatingThreadList = new ArrayList<>();
+
+    // Non private to avoid synthetic class
+    ThreadSafeThreadlessExecutor() {}
+
+    /**
+     * Waits until there is a Runnable, then executes it and all queued Runnables after it.
+     * This, {@link #drain}, and {@link #waitAndDrainWithTimeout(boolean, long, Predicate, Object)}
+     * must only be called by one thread at a time.
+     */
+    public void waitAndDrain() throws InterruptedException {
+      waitAndDrainWithTimeout(true, 0, null, null);
+    }
+
+    /**
+     * Waits for up to specified nanoseconds until there is a Runnable, then executes it and all
+     * queued Runnables after it.
+     *
+     * <p>This method and {@link #drain}, must only be called by one thread at a time.
```

Comment:
Is this still true? If it is true, then we don't need `waitingThreads` and `validatingThreadList` (to be a list), as there could only be one. If this is no longer true (I think the case), then I think `waitingThreads` and `validatingThreadList` isn't working like we'd hope, as `this` lock is held while in `LockSupport.park()`, which means there could only be one. And only one could cause a thread whose Predicate is satisfied to be blocked waiting on a thread (that has the lock) whose Predicate is not satisfied.

I don't think validatingThreadList and waitingThreads being _lists_ is doing much right now, as they seem they'd only have one entry. Also, since we care about Virtual Threads, we don't want Threads to be blocked a long time on an object's monitor (`synchronized (this)`), as that doesn't trigger swapping tasks.

Reply:
Ugh. I need to sit down and read my old comment about the thing we were worried about. I'm still forgetting something about the problem statement :smile:. We are all surprised /s.